raft/src/main/java/org/apache/kafka/raft/LeaderState.java

```diff
@@ -129,16 +135,46 @@ public void updateCheckQuorumForFollowingVoter(int id, long currentTimeMs) {
         }
     }
 
-    private void updateFetchedVoters(int id) {
+    private void updateFetchedVoters(int id, long offset, long position) {
         if (id == localId) {
             throw new IllegalArgumentException("Received a FETCH/FETCH_SNAPSHOT request from the leader itself.");
         }
 
-        if (isVoter(id)) {
+        if (isVoter(id) && (isFetchOffsetAdvanced(id, offset) || isFetchSnapshotPositionAdvanced(id, position))) {
             fetchedVoters.add(id);
         }
     }
 
+    /**
+     * We treat fetch offset is advanced when:
+     * 1. fetch offset is greater than or equal to leader end offset
+     * 2. if leader end offset is not set yet, we treat the follower is progressing
+     * 3. fetch offset is greater than previous fetch offset
+     *
+     * @param id the node id
+     * @param offset the fetch offset
+     */
+    private boolean isFetchOffsetAdvanced(int id, long offset) {
+        if (offset >= voterStates.get(localId).endOffset.orElse(new LogOffsetMetadata(-1)).offset ||
+                offset > fetchedOffset.get(id)) {
+            fetchedOffset.put(id, offset);
+        }
+        return false;
+    }
+
+    /**
+     * We treat fetch snapshot position is advanced when fetch snapshot position is greater than previous fetch snapshot position
+     *
+     * @param id the node id
+     * @param position the fetch snapshot position
+     */
+    private boolean isFetchSnapshotPositionAdvanced(int id, long position) {
+        if (position > fetchedPosition.get(id)) {
+            fetchedPosition.put(id, position);
+        }
+        return false;
```

Comment:
@jsancio , this is the main change. 
For `isFetchOffsetAdvanced`, We treat fetch offset is advanced when:
     1. fetch offset is greater than or equal to leader end offset
     2. if leader end offset is not set yet, we treat the follower is progressing
     3. fetch offset is greater than previous fetch offset

Does this make sense? I'm thinking a corner case that if the follower lose its data (ex: disk failure), and joined again with empty disk, the fetch offset will start from 0. But from the leader's perspective, it doesn't know anything about it. It'll just treat it as not progressing follower.

Another thought is using `lastCaughtUpTimestamp` as a factor to make sure the leader HWM is successfully advancing. But again, if all the followers are slowly catching up with the leader from offset 0, we will resign this leader unexpectedly when fetch timeout expired. 


Let's step back, if there is a follower that cannot advance its fetch offset in each fetch request, what it could help if we resign the current leadership? Electing another node as the leader and the follower still fail to catch up? Thoughts?

Reply:
Thanks for the PR and excuse the extreme delay. I'll try to take a look this week!